### PR TITLE
use async false when load chunk scripts

### PR DESCRIPTION
**Pull request by Markosyan-AR**
#2506 Fix for [#2506](https://github.com/webpack/webpack/issues/2506) bug

1.0 webpack doesnt have async flag and works well with old ie browsers

### CODE REVIEW

File: lib/JsonpMainTemplatePlugin.js [hunk 1]
```diff
@@ -36,7 +36,7 @@ JsonpMainTemplatePlugin.prototype.apply = function(mainTemplate) {
 			"var script = document.createElement('script');",
 			"script.type = 'text/javascript';",
 			"script.charset = 'utf-8';",
-			"script.async = true;",
+			"script.async = false;",
```
Comment (sokra) [inline lib/JsonpMainTemplatePlugin.js]: Is it possible the remove the whole line?

**Comment by Kovensky** [comment]
Not using `async` makes the extra script tag event-loop-blocking.

This probably should be made into an option.

File: lib/JsonpMainTemplatePlugin.js (hunk 1 above)
Comment (Markosyan-AR) [inline lib/JsonpMainTemplatePlugin.js]: by default async flag is true. So no.
hmm, why it is work in v1.0
need additional investigation i think

Comment (sokra) [inline lib/JsonpMainTemplatePlugin.js]: webpack 1 has also `async = true`
https://github.com/webpack/webpack/blob/webpack-1/lib/JsonpMainTemplatePlugin.js#L40

Comment (Markosyan-AR) [inline lib/JsonpMainTemplatePlugin.js]: there it doesn't
https://github.com/webpack/webpack/blob/1.0/lib/JsonpMainTemplate.js#L61

Comment (Markosyan-AR) [inline lib/JsonpMainTemplatePlugin.js]: Noticed that modules, wrapped to functions are executed by IE if they gets from cache
`(function(){
   //code
})()`

chunks are coming like
`window.someFunc()`
and don't execute in that case

Comment (Markosyan-AR) [inline lib/JsonpMainTemplatePlugin.js]: googled a good article about IE loading issues here:
http://www.guypo.com/ies-premature-execution-problem/
and solution is- set src attribute **after** adding element to the dom

**Comment by olaf89** [comment]
Any update on this one?

File: lib/JsonpMainTemplatePlugin.js (hunk 1 above)
Comment (sokra) [inline lib/JsonpMainTemplatePlugin.js]: > and solution is- set src attribute after adding element to the dom
> Also helped to fix this bug

Do you want to change your PR to do this instead?

Comment (Markosyan-AR) [inline lib/JsonpMainTemplatePlugin.js]: [here](https://github.com/webpack/webpack/pull/3400/commits/47290fb2786d02507a812e518c40baf513cea50d) is a change, but it probably break the tests. need to change some tests, don't know which one need to be fixed and how =(

**Comment by r0b1n** [comment]
It actually doesn't make any difference to set `src` after or before appending script element to the DOM.

As I see IE executes **cached** JS **synchronously** at the moment of appending script to DOM (or setting `src`). 
For first page loading it works asynchronously, but second time when JS comes from cache if behaves synchronously (ignores async flag?). 
If cache is turned off it works as expected without any issues..

Current code appends script before setting up Promise array to the `installedChunks` variable. So it is why `jsonp` callback ignores module loading - there is no needed setup in `installedChunks` in case of synchronous script execution.

How this can be solved:
Append script to the DOM only after `installedChunks` is ready to go.
```js
var promise = new Promise(function(resolve, reject) {
    installedChunks[chunkId] = [resolve, reject];
});
installedChunks[chunkId][2] = promise
head.appendChild(script); // append script only if we are ready to process module code.
return promise;
```


Any comments, suggestions?

**Comment by r0b1n** [comment]
@Markosyan-AR will you please check if suggested solution works for you and update this PR?

**Comment by sokra** [comment]
@r0b1n maybe you could just open a new PR?

**Comment by sokra** [comment]
see @r0b1n PR

**Comment by sokra** [comment]
Thanks to @Markosyan-AR for the investigation. :+1:

**Pull request closed without merging**